src/fulfillment/src/main/java/mushop/FulfillmentService.java

```diff
@@ -0,0 +1,101 @@
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+@Singleton
+@Introspected
+public class FulfillmentService {
+    private Connection nc;
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private static final Logger LOG = LoggerFactory.getLogger(FulfillmentService.class);
+    @Value("nats://${mushop.messaging.host}:${mushop.messaging.port}")
+    private String NATS_URL;
+    @Value("${mushop.messaging.subjects.orders}")
+    private String MUSHOP_ORDERS_SUBJECT;
+    @Value("${mushop.messaging.subjects.shipments}")
+    private String MUSHOP_SHIPMENTS_SUBJECT;
+    @Value("${mushop.messaging.simulation-delay}")
+    private Long simulationDelay;
+
+    public FulfillmentService() {
+    }
+
+    @EventListener
+    @Async
+    public void connect(final ServiceStartedEvent event) throws InterruptedException {
+        LOG.info("Connecting to NATS {} and subscribing to subject {}", this.NATS_URL, this.MUSHOP_ORDERS_SUBJECT);
+        boolean connected = false;
+        while (!connected) {
+            try {
+                Future<Boolean> result = Executors.newSingleThreadExecutor().submit(() -> {
+                    try {
+                        nc = Nats.connect(this.NATS_URL);
+                        Dispatcher d = nc.createDispatcher((msg) -> {
+                            try {
+                                handleMessage(msg);
+                            } catch (IOException e) {
+                                e.printStackTrace();
+                            }
+                        });
+                        d.subscribe(this.MUSHOP_ORDERS_SUBJECT);
+                        return Boolean.TRUE;
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                        LOG.error("Connection failed due to {}. Retrying in 5s", e.getMessage());
+                        Thread.sleep(5000l);
+                        return Boolean.FALSE;
+                    }
+                });
+                connected = result.get();
+            } catch (ExecutionException e) {
+                e.printStackTrace();
+            }
+        }
+        LOG.info("Connected to NATS {} and subscribed to subject {}", this.NATS_URL, this.MUSHOP_ORDERS_SUBJECT);
+
+    }
+
+    private void handleMessage(Message message) throws IOException {
+        String response = new String(message.getData(), StandardCharsets.UTF_8);
+        OrderUpdate update = objectMapper.readValue(message.getData(), OrderUpdate.class);
+        LOG.info("got message {} on the mushop orders subject", update);
+        fulfillOrder(update);
+    }
+
+    private void fulfillOrder(OrderUpdate order) {
+        Executors.newSingleThreadExecutor().submit(() -> {
```

Comment:
Does it matter that you're creating this `newSingleThreadExecutor` within another `newSingleThreadExecutor` from the `connect`?

Reply:
No - they are not nested. The call to a SingleThreadExecutor simply creates a new threadpool that can execute async stuff. The one for connect is simply for connection retries.  The dispatcher is setup only if the connection succeeds. 

The dispatcher runs on a separate thread and invokes the callback method. If the callback blocks, then message dispatch slows down. so the call back will process the message on a separate threadpool (since wit will put that thread to sleep to simulate processing delay), making the dispatcher free to handle the next message.

Reply:
However, this prompts me to think in another direction. The singleThreadExecutor is pretty inefficient for this use case in the callback. A cached threadpool is probably a better choice for better message throughput.

Reply:
Also the threadpool that retries the connection should be shutdown after the connection is established to reclaim the thread.
Will fix.

Reply:
Refactored on both Fulfillment as well as the orders application